coding/python_testing.rst

```diff
@@ -42,44 +41,149 @@ The important things to note in this example are:
+
+Running tests standalone
+------------------------
+
+``pySingles`` is an optional argument to the :lmeth:`~lsst.sconsUtils.scripts.BasicSConscript.tests` method that can be used for the rare cases where a test must be run standalone and not with other test files in a shared process.
+
+.. code-block:: python
+
+   scripts.BasicSConscript.tests(pyList=[], pySingles=["testSingle.py"])
+
+The tests are still run using :command:`pytest` but executed one at a time without using any multi-process execution.
+Use of this should be extremely rare.
+In the :lmod:`~lsst.base` package one test file is used to confirm that the LSST import code is working; this can only be tested if we know that it hasn't previously been imported as part of another test.
+The other reason, so far, to run a test standalone is for test classes that dynamically generate large amounts of test data during the set up phase.
+Until it is possible to pin test classes to a particular process with ``pytest-xdist``, tests such as these interact badly when test methods within the class are allocated to different subprocesses since each subprocess will generate the test files.
+This can use significantly more disk and CPU when the test runs, and can even cause Jenkins to fail.
+It is important to ensure that any files listed in ``pySingles`` should be named such that they will not be discovered by `pytest`_.
+The convention is to name these files :file:`test*.py` without the underscore.
+
 .. note::
-   The normal way to run tests is to use :command:`scons` or :command:`scons tests`.
-   :command:`scons` will then execute each of the tests and report any failures.
-   Currently tests are executed one at a time using :command:`python` and checking the command exit status.
-   Eventually :lmod:`~lsst.sconsUtils` will be modified to call :command:`py.test` directly with all Python test files (possibly with a different order each invocation).
-   It is therefore important that during this transition period developers check that tests run correctly with `pytest`_ by explicitly running :command:`py.test` when creating or updating test files.
+  When :command:`pytest` is run by :command:`scons` full warnings are reported, including :py:class:`~exceptions.DeprecationWarning`.
+  Previously these warnings were hidden in the test output but now they are more obvious, allowing you to fix any problems early.
+
+Where does the output go?
+-------------------------
+
+When :command:`scons` runs any tests, the output from those tests is written to the :file:`tests/.tests` directory, and a file is created for each test that is executed.
+If a test command fails, that output is renamed to have a :file:`.failed` extension and is reported by :command:`scons`.
+For the case where :command:`pytest` is running on multiple test files at once, there is a single file created in the :file:`tests/.tests` directory, along with an XML file containing the test output in JUnit format.
```

Comment:
I thought I had asked to move this case (where `pytest` is run on multiple test files) to be first, since it is the normal case now.  Also, is a `.failed` file created in this case?

Reply:
Yes. Sorry. I lost that in the big reorg that @jonathansick asked for. A `.failed` file is always created on failure still, even with the XML file. That was the easiest way to migrate everything.